src/pkg/p2p/preheat/dao/instance/dao.go

```diff
@@ -0,0 +1,112 @@
+package instance
+
+import (
+	"context"
+	"fmt"
+
+	beego_orm "github.com/astaxie/beego/orm"
+	"github.com/goharbor/harbor/src/lib/orm"
+	"github.com/goharbor/harbor/src/lib/q"
+	"github.com/goharbor/harbor/src/pkg/p2p/preheat/models/provider"
+)
+
+// DAO for instance
+type DAO interface {
+	Create(ctx context.Context, instance *provider.Instance) (int64, error)
+	Get(ctx context.Context, id int64) (*provider.Instance, error)
+	Update(ctx context.Context, instance *provider.Instance, props ...string) error
+	Delete(ctx context.Context, id int64) error
+	List(ctx context.Context, query *q.Query) (total int64, ins []*provider.Instance, err error)
+}
+
+// New instance dao
+func New() DAO {
+	return &dao{}
+}
+
+// ListInstanceQuery defines the query params of the instance record.
+type ListInstanceQuery struct {
+	Page     uint
+	PageSize uint
+	Keyword  string
+}
+
+type dao struct{}
+
+var _ DAO = (*dao)(nil)
+
+// Create adds a new distribution instance.
+func (i *dao) Create(ctx context.Context, instance *provider.Instance) (int64, error) {
+	var o, err = orm.FromContext(ctx)
+	if err != nil {
+		return 0, err
+	}
+	return o.Insert(instance)
+}
+
+// Get gets instance from db by id.
+func (i *dao) Get(ctx context.Context, id int64) (*provider.Instance, error) {
+	var o, err = orm.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	di := provider.Instance{ID: id}
+	err = o.Read(&di, "ID")
+	if err == beego_orm.ErrNoRows {
+		return nil, nil
+	}
+	return &di, err
+}
+
+// Update updates distribution instance.
+func (i *dao) Update(ctx context.Context, instance *provider.Instance, props ...string) error {
+	var o, err = orm.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	_, err = o.Update(instance, props...)
+	return err
+}
+
+// Delete deletes one distribution instance by id.
+func (i *dao) Delete(ctx context.Context, id int64) error {
+	var o, err = orm.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	_, err = o.Delete(&provider.Instance{ID: id})
+	return err
+}
+
+// List lists instances by query params.
+func (i *dao) List(ctx context.Context, query *q.Query) (total int64, ins []*provider.Instance, err error) {
+	var o beego_orm.Ormer
+	o, err = orm.FromContext(ctx)
+	if err != nil {
+		return
+	}
+	qt := o.QueryTable(new(provider.Instance))
+
+	if query != nil {
```

Comment:
Here, we should follow the query setter approach used in other modules for keeping consistent.

Reply:
Use "github.com/goharbor/harbor/src/lib/orm".QuerySetter method?

Reply:
yup